File: 2term/jvmWorld/1lab_scala_binaryTree/src/main/scala/Btree.scala
import scala.collection.immutable.Seq

class Node[T](var value: T, var left: Option[Node[T]], var right: Option[Node[T]])

class BinaryTree[T <: CustomTypeTrait] {
  private var root: Option[Node[T]] = None
  def insert(value: T): Unit = {
    root = insertNode(root, value)
  }

  private def insertNode(current: Option[Node[T]], value: T): Option[Node[T]] = current match {
    case None => Some(new Node(value, None, None))
    case Some(node) =>
      if (value.compare(node.value) < 0) {
        node.left = insertNode(node.left, value)
      } else {
        node.right = insertNode(node.right, value)
      }
      Some(node)
  }

  def deleteByValue(value: T): Unit = {
    root = deleteNode(root, value)
  }

  private def deleteNode(current: Option[Node[T]], value: T): Option[Node[T]] = current match {
    case None => None
    case Some(node) =>
      if (value.compare(node.value) < 0) {
        node.left = deleteNode(node.left, value)
      } else if (value.compare(node.value) > 0) {
        node.right = deleteNode(node.right, value)
      } else {
        (node.left, node.right) match {
          case (None, _) => return node.right
          case (_, None) => return node.left
          case _ =>
            val minValue = findMinValue(node.right)
            node.value = minValue
            node.right = deleteNode(node.right, minValue)
        }
      }
      Some(node)
  }

  private def findMinValue(current: Option[Node[T]]): T = current match {
    case Some(node) =>
      node.left match {
        case None => node.value
        case Some(_) => findMinValue(node.left)
      }
    case None => throw new NoSuchElementException("Empty tree")
  }

  def prettyPrint(): Unit = {
    println(s"\n--Binary Tree --")
    printNode(root, "")
  }

  def getParentNode(): Option[Node[T]] = root

  private def printNode(current: Option[Node[T]], indent: String): Unit = current match {
    case Some(node) =>
      printNode(node.left, indent + "    ")
      println(indent + node.value)
      printNode(node.right, indent + "    ")
    case None =>
  }

  def deleteByIndex(index: Int): Unit = {
    root = deleteNodeByIndex(root, index)
  }

  private def deleteNodeByIndex(current: Option[Node[T]], index: Int): Option[Node[T]] = current match {
    case None => None
    case Some(node) =>
      val count = countNodes(node.left)
      if (index < count) {
        node.left = deleteNodeByIndex(node.left, index)
      } else if (index > count) {
        node.right = deleteNodeByIndex(node.right, index - count - 1)
      } else {
        (node.left, node.right) match {
          case (None, _) => return node.right
          case (_, None) => return node.left
          case _ =>
            val minValue = findMinValue(node.right)
            node.value = minValue
            node.right = deleteNodeByIndex(node.right, 0)
        }
      }
      Some(node)
  }

  private def countNodes(current: Option[Node[T]]): Int = current match {
    case Some(node) =>
      1 + countNodes(node.left) + countNodes(node.right)
    case None => 0
  }

  def rebalance(): Unit = {
    val values = inOrderTraversal(root)
    root = buildBalancedTree(values)
  }

  private def inOrderTraversal(current: Option[Node[T]]): List[T] = current match {
    case Some(node) =>
      inOrderTraversal(node.left) ++ List(node.value) ++ inOrderTraversal(node.right)
    case None => Nil
  }

  private def buildBalancedTree(values: List[T]): Option[Node[T]] = values match {
    case Nil => None
    case _ =>
      val mid = values.length / 2
      val node = new Node(values(mid), None, None)
      node.left = buildBalancedTree(values.take(mid))
      node.right = buildBalancedTree(values.drop(mid + 1))
      Some(node)
  }

  def findByValue(value: T): Int = {
    findByValueNode(root, value, 0)
  }

  def forEach(f: T => Unit): Unit = {
    forEachNode(root, f)
  }

  private def forEachNode(current: Option[Node[T]], f: T => Unit): Unit = current match {
    case Some(node) =>
      forEachNode(node.left, f)
      f(node.value)
      forEachNode(node.right, f)
    case None =>
  }

  private def findByValueNode(current: Option[Node[T]], value: T, currentIndex: Int): Int = current match {
    case Some(node) =>
      val leftSubtreeCount = countNodes(node.left)
      if (value.compare(node.value) == 0) {
        currentIndex + leftSubtreeCount
      } else if (value.compare(node.value) < 0) {
        findByValueNode(node.left, value, currentIndex)
      } else {
        findByValueNode(node.right, value, currentIndex + leftSubtreeCount + 1)
      }
    case None => -1
  }

  def depthFirstIterator: Iterator[(T, String)] = new DepthFirstIterator(root)

  def serialize(): String = {
    s"""{"type": "${getType}", "root": ${serializeNode(root)}}"""
  }

  private def serializeNode(node: Option[Node[T]]): String = node match {
    case Some(n) =>
      s"""{"value": "${n.value.serializeToString}", "left": ${serializeNode(n.left)}, "right": ${serializeNode(n.right)}}"""
    case None => "null"
  }

  private def getType: String = {
    if (root.isDefined) root.get.value.typeName else ""
  }

  private class DepthFirstIterator(private val rootNode: Option[Node[T]]) extends Iterator[(T, String)] {
    private val stack: scala.collection.mutable.Stack[(Option[Node[T]], String)] = scala.collection.mutable.Stack()
    private var nextValue: Option[(T, String)] = None

    stack.push((rootNode, "value"))

    override def hasNext: Boolean = {
      if (nextValue.isEmpty) {
        findNext()
      }
      nextValue.isDefined
    }

    override def next(): (T, String) = {
      if (hasNext) {
        val value = nextValue.get
        nextValue = None
        value
      } else {
        throw new NoSuchElementException("No more elements to iterate")
      }
    }

    private def findNext(): Unit = {
      while (stack.nonEmpty && nextValue.isEmpty) {
        val (currentNode, direction) = stack.pop()

        currentNode match {
          case Some(node) =>
            if (node.right.isDefined) {
              stack.push((node.right, "right"))
            }
            if (node.left.isDefined) {
              stack.push((node.left, "left"))
            }
            nextValue = Some((node.value, direction))
          case None =>
        }
      }
    }
  }
}


object BinaryTree {
  def fromList[T <: CustomTypeTrait](values: Seq[T]): BinaryTree[T] = {
    val tree = new BinaryTree[T]
    values.foreach(tree.insert)
    tree
  }

  def fromJsonString[T <: CustomTypeTrait : Manifest](jsonString: String, factory: CustomTypeTrait): BinaryTree[T] = {
    val tree = new BinaryTree[T]
    val parsed = ujson.read(jsonString)
    val rootNode = constructTree(parsed("root"), factory).asInstanceOf[Option[Node[T]]]
    tree.root = rootNode
    tree
  }

  private def constructTree[T <: CustomTypeTrait : Manifest](node: ujson.Value, factory: CustomTypeTrait): Option[Node[CustomTypeTrait]] = {
    if (node.isNull) {
      None
    } else {
      val value = factory.deserializeFromString(node("value").str).get
      val left = constructTree(node("left"), factory)
      val right = constructTree(node("right"), factory)
      Some(new Node[CustomTypeTrait](value, left, right))
    }
  }
}

File: 2term/jvmWorld/2lab_kotlin_btree/src/main/kotlin/Btree.kt
import net.jemzart.jsonkraken.JsonKraken
import net.jemzart.jsonkraken.JsonObject
import net.jemzart.jsonkraken.JsonValue
import net.jemzart.jsonkraken.JsonNull

class BinaryTree<T : CustomTypeInterface> {
    private var root: Node<T>? = null

    fun getParentNode(): Node<T>? {
        return root
    }

    fun insert(value: T) {
        root = insertNode(root, value)
    }

    private fun insertNode(current: Node<T>?, value: T): Node<T> {
        if (current == null) {
            return Node(value)
        }
        if (value.compare(current.value) < 0) {
            current.left = insertNode(current.left, value)
        } else {
            current.right = insertNode(current.right, value)
        }
        return current
    }

    fun deleteByValue(value: T) {
        root = deleteNode(root, value)
    }

    private fun deleteNode(current: Node<T>?, value: T): Node<T>? {
        if (current == null) return null

        if (value.compare(current.value) < 0) {
            current.left = deleteNode(current.left, value)
        } else if (value.compare(current.value) > 0) {
            current.right = deleteNode(current.right, value)
        } else {
            if (current.left == null) return current.right
            if (current.right == null) return current.left

            val minValue = findMinValue(current.right!!)
            current.value = minValue
            current.right = deleteNode(current.right, minValue)
        }
        return current
    }

    private fun findMinValue(current: Node<T>): T {
        return current.left?.let { findMinValue(it) } ?: current.value
    }

    fun prettyPrint() {
        println("\n--Binary Tree --")
        printNode(root, "")
    }

    private fun printNode(current: Node<T>?, indent: String) {
        current?.let {
            printNode(it.left, "$indent    ")
            println("$indent${it.value}")
            printNode(it.right, "$indent    ")
        }
    }

    fun rebalance() {
        val values = inOrderTraversal(root).toList()
        root = buildBalancedTree(values)
    }

    private fun inOrderTraversal(current: Node<T>?): Sequence<T> = sequence {
        if (current == null) return@sequence
        yieldAll(inOrderTraversal(current.left))
        yield(current.value)
        yieldAll(inOrderTraversal(current.right))
    }

    private fun buildBalancedTree(values: List<T>): Node<T>? {
        if (values.isEmpty()) return null
        val mid = values.size / 2
        val node = Node(values[mid])
        node.left = buildBalancedTree(values.take(mid))
        node.right = buildBalancedTree(values.drop(mid + 1))
        return node
    }

    fun findByValue(value: T): Int {
        return findByValueNode(root, value, 0)
    }

    private fun findByValueNode(current: Node<T>?, value: T, currentIndex: Int): Int {
        if (current == null) return -1

        val leftSubtreeCount = countNodes(current.left)
        return when {
            value.compare(current.value) == 0 -> currentIndex + leftSubtreeCount
            value.compare(current.value) < 0 -> findByValueNode(current.left, value, currentIndex)
            else -> findByValueNode(current.right, value, currentIndex + leftSubtreeCount + 1)
        }
    }

    private fun countNodes(current: Node<T>?): Int {
        if (current == null) return 0
        return 1 + countNodes(current.left) + countNodes(current.right)
    }

    fun forEach(action: (T) -> Unit) {
        forEachNode(root, action)
    }

    private fun forEachNode(current: Node<T>?, action: (T) -> Unit) {
        current?.let {
            forEachNode(it.left, action)
            action(it.value)
            forEachNode(it.right, action)
        }
    }

    fun serialize(): String {
        return """{"type": "${getType()}", "root": ${serializeNode(root)}}"""
    }

    private fun serializeNode(node: Node<T>?): String {
        return node?.let {
            """{"value": "${it.value.serializeToString()}", "left": ${serializeNode(it.left)}, "right": ${serializeNode(it.right)}}"""
        } ?: "null"
    }

    private fun getType(): String {
        return root?.value?.typeName ?: ""
    }

    private inner class DepthFirstIterator(private val rootNode: Node<T>?) : Iterator<Pair<T, String>> {
        private val stack: MutableList<Pair<Node<T>?, String>> = mutableListOf()
        private var nextValue: Pair<T, String>? = null

        init {
            stack.add(Pair(rootNode, "value"))
        }

        override fun hasNext(): Boolean {
            if (nextValue == null) {
                findNext()
            }
            return nextValue != null
        }

        override fun next(): Pair<T, String> {
            if (hasNext()) {
                val value = nextValue!!
                nextValue = null
                return value
            } else {
                throw NoSuchElementException("No more elements to iterate")
            }
        }

        private fun findNext() {
            while (stack.isNotEmpty() && nextValue == null) {
                val (currentNode, direction) = stack.removeAt(stack.size - 1)

                currentNode?.let {
                    it.right?.let { rightNode -> stack.add(Pair(rightNode, "right")) }
                    it.left?.let { leftNode -> stack.add(Pair(leftNode, "left")) }
                    nextValue = Pair(it.value, direction)
                }
            }
        }
    }



    companion object {
        fun <T : CustomTypeInterface> fromList(values: List<T>): BinaryTree<T> {
            val tree = BinaryTree<T>()
            values.forEach { tree.insert(it) }
            return tree
        }

        fun <T : CustomTypeInterface> fromJsonString(jsonString: String, factory: CustomTypeInterface): BinaryTree<T> {
            val tree = BinaryTree<T>()
            val parsed = JsonKraken.deserialize(jsonString) as JsonObject
            val rootNode = constructTree(parsed["root"], factory) as Node<T>?
            tree.root = rootNode
            return tree
        }

        private fun <T : CustomTypeInterface> constructTree(node: JsonValue, factory: T): Node<T>? {
            if (node is JsonNull) {
                return null
            } else {
                val nodeObj = node as JsonObject
                val value = factory.deserializeFromString(nodeObj["value"].cast<String>()) as T? ?: return null
                val left = constructTree(nodeObj["left"], factory)
                val right = constructTree(nodeObj["right"], factory)
                return Node(value, left, right)
            }
        }
    }
}